cleverhans/attacks_tf.py

```diff
@@ -725,9 +757,162 @@ def compare(x, y):
+                              relative importance of size of the pururbation
+                              and confidence of classification.
+                              A smaller value of this constant gives lower
+                              distortion results.
+        :param largest_const: When the tradeoff-constant exceeds this value,
+                              the attack terminates. Larger values gives lower
+                              distortion results.
+        :param const_factor: How much to increase the tradeoff-constant by
+                             on each iteration of the attack if the prior
+                             iteration failed.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        self.initial_const = initial_const
+        self.largest_const = largest_const
+        self.const_factor = const_factor
+        self.l2_attack = CarliniWagnerL2(sess, model, 1, confidence,
+                                         targeted, learning_rate,
+                                         1, max_iterations,
+                                         abort_early, initial_const,
+                                         clip_min, clip_max, num_labels, shape,
+                                         extension=0)
+
+    def attack(self, instances, targets):
+        """
+        Perform the L_0 attack on the given instance for the given targets.
+
+        If self.targeted is true, then the targets represents the target labels
+        If self.targeted is false, then targets are the original class labels
+        """
+
+        r = []
+        for i, (instance, target) in enumerate(zip(instances, targets)):
+            _logger.debug(("Running CWL0 attack on instance " +
+                           "{} of {}").format(i, len(instances)))
+            r.extend(self.attack_single(instance, target))
+        return np.array(r)
+
+    def attack_single(self, instance, target):
+        """
+        Run the attack on a single instance and label.
+        """
+
+        # the pixels we can change
+        valid = np.ones(instance.shape)
+
+        # the previous image
+        prev = np.copy([instance])
+
+        last_solution = [instance]
+        const = self.initial_const
+        equal_count = 0
+
+        while True:
+            # try to solve given this valid map
+            self.l2_attack.source_image = np.copy(prev)
+            while const < self.largest_const:
+                # try solving for each value of the constant
+                self.l2_attack.initial_const = const
+                res = self.l2_attack.attack_batch(np.copy([instance]),
+                                                  np.array([target]),
+                                                  mask=np.array([valid]))
+                if res is not None:
+                    break
+                const *= self.const_factor
```

Comment:
Just double checking that we dont' want to initialize const to self.initial_const before entering this inner loop.

Reply:
Nope, want it to stay at the old value.